From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nds-ports <nds-ports@localhost>
Date: Tue, 6 Oct 2026 07:09:59 +0200
Subject: [PATCH] Parse text without scanf and strtod

Savegames, their comments for the menu and point files are read a word
at a time with COM_ReadWord and Q_atof instead of fscanf, entity fields
with Q_atof instead of atof. The C library's scanf and strtod, with
their wide character tables, take more room than all of Quake's menus.
---
 WinQuake/common.c   | 22 ++++++++++++++++++++++
 WinQuake/common.h   |  1 +
 WinQuake/host_cmd.c | 14 +++++++-------
 WinQuake/menu.c     |  5 ++---
 WinQuake/pr_edict.c |  4 ++--
 WinQuake/r_part.c   |  3 ++-
 6 files changed, 36 insertions(+), 13 deletions(-)

diff --git a/WinQuake/common.c b/WinQuake/common.c
index 6da0f96..c1927ae 100644
--- a/WinQuake/common.c
+++ b/WinQuake/common.c
@@ -901,6 +901,28 @@ void COM_DefaultExtension (char *path, char *extension)
 }
 
 
+/*
+==============
+COM_ReadWord
+
+Reads the next word of a text file, like the %s of fscanf, the word is
+empty at the end of the file
+==============
+*/
+char *COM_ReadWord (FILE *f, char *word, int size)
+{
+	int		c, len;
+
+	do
+		c = fgetc (f);
+	while (c != EOF && c <= ' ');
+	for (len = 0 ; c != EOF && c > ' ' ; c = fgetc (f))
+		if (len < size - 1)
+			word[len++] = c;
+	word[len] = 0;
+	return word;
+}
+
 /*
 ==============
 COM_Parse
diff --git a/WinQuake/common.h b/WinQuake/common.h
index 87d4a55..254d752 100644
--- a/WinQuake/common.h
+++ b/WinQuake/common.h
@@ -142,6 +142,7 @@ extern	char		com_token[1024];
 extern	qboolean	com_eof;
 
 char *COM_Parse (char *data);
+char *COM_ReadWord (FILE *f, char *word, int size);
 
 
 extern	int		com_argc;
diff --git a/WinQuake/host_cmd.c b/WinQuake/host_cmd.c
index 8b76bfe..a809215 100644
--- a/WinQuake/host_cmd.c
+++ b/WinQuake/host_cmd.c
@@ -614,7 +614,7 @@ void Host_Loadgame_f (void)
 		return;
 	}
 
-	fscanf (f, "%i\n", &version);
+	version = atoi (COM_ReadWord (f, str, sizeof(str)));
 	if (version != SAVEGAME_VERSION)
 	{
 		fclose (f);
@@ -622,11 +622,11 @@ void Host_Loadgame_f (void)
 		SCR_EndLoadingPlaque ();
 		return;
 	}
-	fscanf (f, "%s\n", str);
+	COM_ReadWord (f, str, sizeof(str));
 	for (i=0 ; i<NUM_SPAWN_PARMS ; i++)
-		fscanf (f, "%f\n", &spawn_parms[i]);
+		spawn_parms[i] = Q_atof (COM_ReadWord (f, str, sizeof(str)));
 // this silliness is so we can load 1.06 save files, which have float skill values
-	fscanf (f, "%f\n", &tfloat);
+	tfloat = Q_atof (COM_ReadWord (f, str, sizeof(str)));
 	current_skill = (int)(tfloat + 0.1);
 	Cvar_SetValue ("skill", (float)current_skill);
 
@@ -636,8 +636,8 @@ void Host_Loadgame_f (void)
 	Cvar_SetValue ("teamplay", 0);
 #endif
 
-	fscanf (f, "%s\n",mapname);
-	fscanf (f, "%f\n",&time);
+	COM_ReadWord (f, mapname, sizeof(mapname));
+	time = Q_atof (COM_ReadWord (f, str, sizeof(str)));
 
 	CL_Disconnect_f ();
 	
@@ -659,7 +659,7 @@ void Host_Loadgame_f (void)
 
 	for (i=0 ; i<MAX_LIGHTSTYLES ; i++)
 	{
-		fscanf (f, "%s\n", str);
+		COM_ReadWord (f, str, sizeof(str));
 		sv.lightstyles[i] = Hunk_Alloc (strlen(str)+1);
 		strcpy (sv.lightstyles[i], str);
 	}
diff --git a/WinQuake/menu.c b/WinQuake/menu.c
index b5c0be2..9a72ecc 100644
--- a/WinQuake/menu.c
+++ b/WinQuake/menu.c
@@ -475,7 +475,6 @@ void M_ScanSaves (void)
 	int		i, j;
 	char	name[MAX_OSPATH];
 	FILE	*f;
-	int		version;
 
 	for (i=0 ; i<MAX_SAVEGAMES ; i++)
 	{
@@ -485,8 +484,8 @@ void M_ScanSaves (void)
 		f = fopen (name, "r");
 		if (!f)
 			continue;
-		fscanf (f, "%i\n", &version);
-		fscanf (f, "%79s\n", name);
+		COM_ReadWord (f, name, sizeof(name));	// the version
+		COM_ReadWord (f, name, sizeof(name));
 		strncpy (m_filenames[i], name, sizeof(m_filenames[i])-1);
 
 	// change _ back to space
diff --git a/WinQuake/pr_edict.c b/WinQuake/pr_edict.c
index f958446..5d2acb1 100644
--- a/WinQuake/pr_edict.c
+++ b/WinQuake/pr_edict.c
@@ -743,7 +743,7 @@ qboolean	ED_ParseEpair (void *base, ddef_t *key, char *s)
 		break;
 		
 	case ev_float:
-		*(float *)d = atof (s);
+		*(float *)d = Q_atof (s);
 		break;
 		
 	case ev_vector:
@@ -755,7 +755,7 @@ qboolean	ED_ParseEpair (void *base, ddef_t *key, char *s)
 			while (*v && *v != ' ')
 				v++;
 			*v = 0;
-			((float *)d)[i] = atof (w);
+			((float *)d)[i] = Q_atof (w);
 			w = v = v+1;
 		}
 		break;
diff --git a/WinQuake/r_part.c b/WinQuake/r_part.c
index f949418..59842d8 100644
--- a/WinQuake/r_part.c
+++ b/WinQuake/r_part.c
@@ -215,7 +215,8 @@ void R_ReadPointFile_f (void)
 	c = 0;
 	for ( ;; )
 	{
-		r = fscanf (f,"%f %f %f\n", &org[0], &org[1], &org[2]);
+		for (r = 0 ; r < 3 && *COM_ReadWord (f, name, sizeof(name)) ; r++)
+			org[r] = Q_atof (name);
 		if (r != 3)
 			break;
 		c++;
